File: backend/services/platformService.js
const axios = require('axios');
const cheerio = require('cheerio');

// --- LeetCode ---
const fetchLeetCode = async (handle) => {
    console.log(`Fetching LeetCode for: ${handle}`);
    try {
        const query = `
            query getUserData($username: String!) {
                matchedUser(username: $username) {
                    profile {
                        ranking
                        reputation
                    }
                    submitStats {
                        acSubmissionNum {
                            difficulty
                            count
                        }
                    }
                }
                userContestRanking(username: $username) {
                    rating
                    globalRanking
                }
                userContestRankingHistory(username: $username) {
                    attended
                    rating
                    contest {
                        startTime
                    }
                }
            }
        `;
        const response = await axios.post('https://leetcode.com/graphql', {
            query,
            variables: { username: handle }
        });

        // Fetch calendar separately to avoid complexity in a single query if needed, 
        // or just add it to the main query. Let's add it to the main query.
        const calendarQuery = `
            query userProfileCalendar($username: String!) {
                matchedUser(username: $username) {
                    userCalendar {
                        submissionCalendar
                        streak
                        totalActiveDays
                    }
                }
            }
        `;
        const calendarResponse = await axios.post('https://leetcode.com/graphql', {
            query: calendarQuery,
            variables: { username: handle }
        });

        if (response.data.errors || !response.data.data.matchedUser) {
            return { platform: 'LeetCode', handle, error: "User not found", success: false };
        }

        const data = response.data.data;
        const solved = data.matchedUser.submitStats.acSubmissionNum.find(s => s.difficulty === 'All').count;
        const rating = data.userContestRanking ? Math.round(data.userContestRanking.rating) : null;

        // History: only attended contests
        const history = data.userContestRankingHistory
            ? data.userContestRankingHistory
                .filter(x => x.attended)
                .map(x => ({
                    rating: Math.round(x.rating),
                    date: new Date(x.contest.startTime * 1000).toISOString().split('T')[0]
                }))
            : [];

        const calendarData = calendarResponse.data.data?.matchedUser?.userCalendar || {};
        const submissionCalendar = calendarData.submissionCalendar ? JSON.parse(calendarData.submissionCalendar) : {};
        
        // Convert submissionCalendar (timestamp in seconds) to YYYY-MM-DD
        const calendar = {};
        Object.entries(submissionCalendar).forEach(([ts, count]) => {
            const date = new Date(parseInt(ts) * 1000).toISOString().split('T')[0];
            calendar[date] = (calendar[date] || 0) + count;
        });

        return {
            platform: 'LeetCode',
            handle,
            solved,
            rating,
            ranking: data.matchedUser.profile.ranking,
            history,
            calendar,
            streak: calendarData.streak,
            totalActiveDays: calendarData.totalActiveDays,
            success: true
        };
    } catch (err) {
        console.error(`LeetCode fetch error for ${handle}:`, err.message);
        return { platform: 'LeetCode', handle, error: "Failed to fetch", success: false };
    }
};

// --- Codeforces ---
const fetchCodeforces = async (handle) => {
    console.log(`Fetching Codeforces for: ${handle}`);
    try {
        // Parallel calls: Info and Rating History
        const [infoRes, ratingRes] = await Promise.all([
            axios.get(`https://codeforces.com/api/user.info?handles=${handle}`),
            axios.get(`https://codeforces.com/api/user.rating?handle=${handle}`)
        ]);

        if (infoRes.data.status !== 'OK') return null;

        const user = infoRes.data.result[0];
        const history = ratingRes.data.status === 'OK'
            ? ratingRes.data.result.map(r => ({
                rating: r.newRating,
                date: new Date(r.ratingUpdateTimeSeconds * 1000).toISOString().split('T')[0]
            }))
            : [];

        // Fetch recent submissions for activity and TOTAL SOLVED
        let calendar = {};
        let solved = 0;
        try {
            const statusRes = await axios.get(`https://codeforces.com/api/user.status?handle=${handle}&from=1&count=10000`);
            if (statusRes.data.status === 'OK') {
                const uniqueSolved = new Set();
                statusRes.data.result.forEach(sub => {
                    const date = new Date(sub.creationTimeSeconds * 1000).toISOString().split('T')[0];
                    if (sub.verdict === 'OK') {
                        calendar[date] = (calendar[date] || 0) + 1;
                        // Unique problem key: contestId + index (e.g., 1234A)
                        uniqueSolved.add(`${sub.problem.contestId}${sub.problem.index}`);
                    }
                });
                solved = uniqueSolved.size;
            }
        } catch (e) { console.error("CF Status fetch error", e.message); }

        return {
            platform: 'Codeforces',
            handle,
            rating: user.rating,
            rank: user.rank,
            maxRating: user.maxRating,
            solved,
            history,
            calendar,
            success: true
        };
    } catch (err) {
        console.error(`Codeforces fetch error for ${handle}:`, err.message);
        return { platform: 'Codeforces', handle, error: "Failed to fetch", success: false };
    }
};

// --- GFG (GeeksforGeeks) ---
const fetchGFG = async (handle) => {
    console.log(`Fetching GFG for: ${handle}`);
    try {
        const response = await axios.get(`https://www.geeksforgeeks.org/user/${handle}/`, {
            headers: { 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36' }
        });
        const $ = cheerio.load(response.data);
        
        // GFG stats are often in spans with specific classes
        const solvedText = $('.scoreCard_head_left--score__39_Z0').first().text() || 
                           $('.tabs_main_content .solve-count').first().text() || "0";
        const solved = parseInt(solvedText.replace(/\D/g, '')) || 0;
        
        return {
            platform: 'GFG',
            handle,
            solved,
            calendar: {},
            history: [],
            success: true
        };
    } catch (err) {
        console.error(`GFG fetch error for ${handle}:`, err.message);
        return { platform: 'GFG', handle, error: "Failed to fetch", success: false };
    }
};

// --- AtCoder ---
const fetchAtCoder = async (handle) => {
    console.log(`Fetching AtCoder for: ${handle}`);
    try {
        const [profileRes, historyRes] = await Promise.all([
            axios.get(`https://atcoder.jp/users/${handle}`, { headers: { 'User-Agent': 'Mozilla/5.0' } }),
            axios.get(`https://atcoder.jp/users/${handle}/history/json`, { headers: { 'User-Agent': 'Mozilla/5.0' } })
        ]);

        const $ = cheerio.load(profileRes.data);
        let rating = 0;
        let solved = 0;
        
        $('th').each((i, el) => {
            const text = $(el).text().trim();
            if (text === 'Rating') {
                const val = $(el).next('td').text().trim();
                const match = val.match(/^(\d+)/);
                if (match) rating = parseInt(match[1]);
            }
        });

        // Scrape solved count if possible (AtCoder doesn't show it directly on main profile always, 
        // but often in "Accepted" under contest results or via a separate link. 
        // For now, let's look for any 'Accepted' text)
        const acceptedMatch = profileRes.data.match(/Accepted<\/th><td>(\d+)<\/td>/);
        if (acceptedMatch) solved = parseInt(acceptedMatch[1]);

        const history = historyRes.data.map(x => ({
            rating: x.NewRating,
            date: new Date(x.EndTime).toISOString().split('T')[0]
        }));

        return {
            platform: 'AtCoder',
            handle,
            rating,
            solved,
            history,
            success: true
        };
    } catch (err) {
        console.error(`AtCoder fetch error for ${handle}:`, err.message);
        return { platform: 'AtCoder', handle, error: "Failed to fetch", success: false };
    }
};

// --- CodeChef ---
const fetchCodeChef = async (handle) => {
    console.log(`Fetching CodeChef for: ${handle}`);
    try {
        const response = await axios.get(`https://www.codechef.com/users/${handle}`, {
            headers: { 'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36' }
        });
        const $ = cheerio.load(response.data);
        const rating = $('.rating-number').first().text().replace(/\D/g, '');
        const stars = $('.rating-star').text().trim();

        // Extract History from Script Tag (CodeChef embeds it in a Highcharts config)
        let history = [];
        const scripts = $('script');
        scripts.each((i, s) => {
            const content = $(s).html();
            if (content && content.includes('all_rating')) {
                // Regex to find the rating history array
                const match = content.match(/all_rating\s*=\s*(\[.*?\]);/);
                if (match) {
                    try {
                        const rawHistory = JSON.parse(match[1]);
                        history = rawHistory.map(item => ({
                            rating: parseInt(item.rating),
                            date: `${item.year}-${item.month.padStart(2, '0')}-${item.day.padStart(2, '0')}`
                        }));
                    } catch (e) {
                        console.error("CodeChef history parse error", e.message);
                    }
                }
            }
        });

        return {
            platform: 'CodeChef',
            handle,
            rating: parseInt(rating) || 0,
            stars: stars || '0',
            solved: 0, // CodeChef solved count is in another complex section
            history: history,
            success: !!rating
        };
    } catch (err) {
        console.error(`CodeChef fetch error for ${handle}:`, err.message);
        return { platform: 'CodeChef', handle, error: "Failed to fetch", success: false };
    }
};

module.exports = { fetchLeetCode, fetchCodeforces, fetchCodeChef, fetchAtCoder, fetchGFG };

File: flask_based_app/services/platform_service.py
import requests
from bs4 import BeautifulSoup
from datetime import datetime
import json
import re
import logging

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# LeetCode — uses the alfa-leetcode-api proxy (much more reliable than
#             hitting the GraphQL endpoint directly, which now blocks
#             server-side requests without a browser session).
# ---------------------------------------------------------------------------
LEETCODE_PROXY = 'https://alfa-leetcode-api.onrender.com'

def fetch_leetcode(handle):
    logger.info(f"Fetching LeetCode for: {handle}")
    try:
        # Profile + solved stats from the proxy
        profile_res = requests.get(f"{LEETCODE_PROXY}/{handle}", timeout=15)
        solved_res  = requests.get(f"{LEETCODE_PROXY}/{handle}/solved", timeout=15)

        if profile_res.status_code != 200:
            return {'platform': 'LeetCode', 'handle': handle, 'error': 'User not found', 'success': False}

        profile = profile_res.json()

        # The proxy returns {"username": ..., "ranking": ..., ...}
        if not profile.get('username'):
            return {'platform': 'LeetCode', 'handle': handle, 'error': 'User not found', 'success': False}

        solved = 0
        rating = None

        if solved_res.status_code == 200:
            solved_data = solved_res.json()
            solved = solved_data.get('solvedProblem', 0)

        # Try to get contest rating and history (may be empty if user hasn't participated)
        history = []
        try:
            contest_res = requests.get(f"{LEETCODE_PROXY}/{handle}/contest", timeout=10)
            if contest_res.status_code == 200:
                contest_data = contest_res.json()
                # contestRating field or extract from contestParticipation
                if contest_data.get('contestRating'):
                    rating = round(contest_data['contestRating'])

                # Build history from contestParticipation
                participations = contest_data.get('contestParticipation', [])
                for p in participations:
                    if p.get('rating'):
                        entry = {'rating': round(p['rating'])}
                        # Try to extract date from contest title or timestamp
                        if p.get('contest') and p['contest'].get('startTime'):
                            entry['date'] = datetime.fromtimestamp(int(p['contest']['startTime'])).strftime('%Y-%m-%d')
                        elif p.get('startTime'):
                            entry['date'] = datetime.fromtimestamp(int(p['startTime'])).strftime('%Y-%m-%d')
                        else:
                            entry['date'] = ''
                        history.append(entry)

                if not rating and participations:
                    last = participations[-1]
                    if last.get('rating'):
                        rating = round(last['rating'])
        except Exception:
            pass

        return {
            'platform': 'LeetCode',
            'handle': handle,
            'solved': solved,
            'rating': rating,
            'ranking': profile.get('ranking'),
            'history': history,
            'success': True
        }
    except Exception as e:
        logger.error(f"LeetCode error: {e}")
        return {'platform': 'LeetCode', 'handle': handle, 'error': f'Failed to fetch: {e}', 'success': False}


# ---------------------------------------------------------------------------
# Codeforces — official API, works fine
# ---------------------------------------------------------------------------
def fetch_codeforces(handle):
    logger.info(f"Fetching Codeforces for: {handle}")
    try:
        info_res = requests.get(f"https://codeforces.com/api/user.info?handles={handle}", timeout=10)
        rating_res = requests.get(f"https://codeforces.com/api/user.rating?handle={handle}", timeout=10)
        
        if info_res.json().get('status') != 'OK':
            return {'platform': 'Codeforces', 'handle': handle, 'error': 'User not found', 'success': False}
            
        user = info_res.json()['result'][0]
        history = []
        if rating_res.json().get('status') == 'OK':
            history = [{'rating': r['newRating'], 'date': datetime.fromtimestamp(r['ratingUpdateTimeSeconds']).strftime('%Y-%m-%d')}
                       for r in rating_res.json()['result']]
                       
        solved = 0
        try:
            status_res = requests.get(f"https://codeforces.com/api/user.status?handle={handle}&from=1&count=1000", timeout=10)
            if status_res.json().get('status') == 'OK':
                unique_solved = set()
                for sub in status_res.json()['result']:
                    if sub.get('verdict') == 'OK':
                        unique_solved.add(f"{sub['problem'].get('contestId')}{sub['problem'].get('index')}")
                solved = len(unique_solved)
        except Exception:
            pass

        return {
            'platform': 'Codeforces', 'handle': handle, 'rating': user.get('rating'),
            'maxRating': user.get('maxRating'), 'solved': solved, 'history': history, 'success': True
        }
    except Exception as e:
        logger.error(f"Codeforces error: {e}")
        return {'platform': 'Codeforces', 'handle': handle, 'error': f'Failed to fetch: {e}', 'success': False}


# ---------------------------------------------------------------------------
# CodeChef — uses codechef-api.vercel.app community proxy.
#             Falls back to direct scraping if the proxy is down.
# ---------------------------------------------------------------------------
def fetch_codechef(handle):
    logger.info(f"Fetching CodeChef for: {handle}")

    # --- Attempt 1: Community proxy API ---
    try:
        proxy_res = requests.get(f"https://codechef-api.vercel.app/handle/{handle}", timeout=12)
        if proxy_res.status_code == 200:
            data = proxy_res.json()
            if data.get('success') != False and (data.get('currentRating') or data.get('rating')):
                rating = data.get('currentRating') or data.get('rating')
                return {
                    'platform': 'CodeChef',
                    'handle': handle,
                    'rating': int(rating) if rating else None,
                    'solved': data.get('totalProblemsSolved') or data.get('problemsSolved'),
                    'success': True
                }
    except Exception as e:
        logger.warning(f"CodeChef proxy failed, trying direct scrape: {e}")

    # --- Attempt 2: Direct scrape ---
    try:
        headers = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'}
        res = requests.get(f"https://www.codechef.com/users/{handle}", headers=headers, timeout=10)
        soup = BeautifulSoup(res.text, 'html.parser')

        # Try multiple known selectors (CodeChef changes their layout often)
        rating = None
        for selector in ['.rating-number', '.rating-header .rating', '.user-details-container .rating']:
            el = soup.select_one(selector)
            if el:
                digits = ''.join(filter(str.isdigit, el.text))
                if digits:
                    rating = int(digits)
                    break

        if rating is None:
            return {'platform': 'CodeChef', 'handle': handle, 'error': 'Could not parse rating', 'success': False}

        return {'platform': 'CodeChef', 'handle': handle, 'rating': rating, 'success': True}
    except Exception as e:
        logger.error(f"CodeChef scrape error: {e}")
        return {'platform': 'CodeChef', 'handle': handle, 'error': f'Failed to fetch: {e}', 'success': False}


# ---------------------------------------------------------------------------
# AtCoder — direct scrape, works reliably
# ---------------------------------------------------------------------------
def fetch_atcoder(handle):
    logger.info(f"Fetching AtCoder for: {handle}")
    try:
        headers = {'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'}
        res = requests.get(f"https://atcoder.jp/users/{handle}", headers=headers, timeout=10)
        soup = BeautifulSoup(res.text, 'html.parser')
        rating = 0
        for th in soup.find_all('th'):
            if th.text.strip() == 'Rating':
                td = th.find_next_sibling('td')
                if td:
                    match = re.match(r'^(\d+)', td.text.strip())
                    if match:
                        rating = int(match.group(1))
        return {'platform': 'AtCoder', 'handle': handle, 'rating': rating, 'success': True}
    except Exception as e:
        logger.error(f"AtCoder error: {e}")
        return {'platform': 'AtCoder', 'handle': handle, 'error': f'Failed to fetch: {e}', 'success': False}


# ---------------------------------------------------------------------------
# GFG — uses community proxy APIs.
#        Direct scraping doesn't work because GFG is JS-rendered.
# ---------------------------------------------------------------------------
def fetch_gfg(handle):
    logger.info(f"Fetching GFG for: {handle}")

    # --- Attempt 1: Community stats API ---
    try:
        stats_res = requests.get(
            "https://geeks-for-geeks-stats-api.vercel.app/",
            params={'raw': 'Y', 'userName': handle},
            timeout=15
        )
        if stats_res.status_code == 200:
            data = stats_res.json()
            # Validate: must have recognizable fields and a sane number
            total = data.get('totalProblemsSolved')
            if total is not None and isinstance(total, (int, float)) and 0 <= total < 10000:
                return {
                    'platform': 'GFG',
                    'handle': handle,
                    'solved': int(total),
                    'success': True
                }
    except Exception as e:
        logger.warning(f"GFG stats API failed: {e}")

    # --- Attempt 2: Alternate community API ---
    try:
        alt_res = requests.get(f"https://gfg-api-fefa.onrender.com/{handle}", timeout=15)
        if alt_res.status_code == 200:
            data = alt_res.json()
            total = data.get('totalProblemsSolved')
            if total is not None and isinstance(total, (int, float, str)):
                try:
                    total_int = int(total)
                except (ValueError, TypeError):
                    total_int = 0
                if 0 <= total_int < 10000:
                    return {
                        'platform': 'GFG',
                        'handle': handle,
                        'solved': total_int,
                        'success': True
                    }
    except Exception as e:
        logger.warning(f"GFG alternate API failed: {e}")

    # If both APIs failed, return a clean error — no scraping (GFG is JS-rendered)
    return {
        'platform': 'GFG',
        'handle': handle,
        'error': 'Could not fetch stats — check your handle',
        'success': False
    }


# ---------------------------------------------------------------------------
# Main dispatcher
# ---------------------------------------------------------------------------
def fetch_stats(handles):
    results = []
    for item in handles:
        platform = item.get('platform')
        handle = item.get('handle')
        if not handle:
            continue
        if platform == 'LeetCode':
            results.append(fetch_leetcode(handle))
        elif platform == 'Codeforces':
            results.append(fetch_codeforces(handle))
        elif platform == 'CodeChef':
            results.append(fetch_codechef(handle))
        elif platform == 'AtCoder':
            results.append(fetch_atcoder(handle))
        elif platform == 'GFG':
            results.append(fetch_gfg(handle))
    return results
